invoices/update.go

```diff
@@ -206,15 +229,91 @@ func updateMpp(ctx *invoiceUpdateCtx,
 		return &update, ctx.acceptRes(resultAccepted), nil
 	}
 
+	var (
```

Comment:
If we add a check for an invalid AMP payload, _before_ the the check for the total amount, then it would be possible to send a corrupt share to cancel all HTLCs w/ that set ID. This lets the sender bail out if they know things won't work for w/e reason. It would also be useful for more iterative probing.

Reply:
it isn't really possible to send an invalid AMP payload, and it's only possible to know if the shares are invalid after receiving all of them. if the receiver tries to reconstruct early they will always be invalid. given that the amount _is_ the termination condition, i don't think there's any getting around that.

the current iteration does fail back all HTLCs if the reconstruction fails tho, to support the probing use case.

Reply:
Hmm, yeah I get what you mean. Perhaps this can be done by extending the `ValidateParsedPayloadTypes` checks: if the sender sends a payload w/o an amount, but with a valid set ID of already held HTLCs, they call get cancelled back? 

>  given that the amount is the termination condition, i don't think there's any getting around that.

Another idea: we reserve a special value of the `childIndex` that can be used to signal intent to bail the payment early. 

The motivation here is to support a fast-fail mode w/o having to wait for the normal MPP timeout. This can be useful in contexts like attempting to Loop In wherein the user doesn't really know how much they can Loop In, and even if we fail w/ probing the full amount, if can get some additional feedback, then we can offer a: "Loop In as much as you can" mode. 

Stepping back a bit, perhaps this can just be implemented by moving one layer higher: a special custom record that just says "ok, can you guys cancel things back now". Applications would be able to use the to-be-designed API for handling hodl AMP payments to handle this, or we'd grow to do it automatically.

Reply:
Updated the PR to include immediate cancellation when seed reconstruction fails as well as some tests.

> Stepping back a bit, perhaps this can just be implemented by moving one layer higher: a special custom record that just says "ok, can you guys cancel things back now". Applications would be able to use the to-be-designed API for handling hodl AMP payments to handle this, or we'd grow to do it automatically.

This idea sounds pretty workable! Seems like a pretty contained change as well, also wouldn't need reserved values, special cases, etc in the base protocol

Reply:
So the receiver gets the cancel signal and cancels all htlcs back, but does the sender learn from that which htlcs were actually held by the receiver vs the ones that were still in-flight or didn't even reach the receiver? I suppose the sender can inspect the failure message to see what the origin of the failure is. Only when an htlc arrives at the receiver after the set has already been cancelled, and is then cancelled back, what does this look like to the sender? Will it be clear that that final htlc wasn't part of the set and reached the receiver "independently", possibly reusing liquidity that was just released?